Preparatory work for the unit loading rework

keszybz opened the pull request:
This isn't ready yet. The preparatory parts should be OK, but the last three patches are not fully finished. I'm posting this to get some feedback about the general direction. If this is accepted, I'll do the polishing.

poettering commented on src/basic/path-util.h:
```diff
@@ -71,16 +71,22 @@ static inline bool path_equal_ptr(const char *a, const char *b) {
+        }
+
+        return NULL;
+}
```
i think this is long enough to just be a regular function

poettering commented on src/basic/path-util.h:
```diff
@@ -71,16 +71,22 @@ static inline bool path_equal_ptr(const char *a, const char *b) {
+#define PATH_STARTSWITH_SET(p, ...)                               \
+        ({                                                        \
+                const char *_p = (p);                             \
+                path_startswith_strv(_p, STRV_MAKE(__VA_ARGS__)); \
```
why the _p? that can be dropped so that this is just a regular function call

poettering commented on src/basic/unit-name.c:
```diff
@@ -142,19 +142,18 @@ int unit_name_to_prefix(const char *n, char **ret) {
 
 int unit_name_to_instance(const char *n, char **instance) {
```
maybe rename 'instance' to 'ret' to follow our usual style while we are at it

poettering commented on src/basic/unit-name.h:
```diff
@@ -10,9 +10,9 @@
-        UNIT_NAME_ANY = UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE|UNIT_NAME_TEMPLATE,
+        UNIT_NAME_TEMPLATE = 1 << 1, /* Allow foo@.service */
+        UNIT_NAME_INSTANCE = 1 << 2, /* Allow foo@bar.service */
+        UNIT_NAME_ANY = UNIT_NAME_PLAIN|UNIT_NAME_TEMPLATE|UNIT_NAME_INSTANCE,
```
why this reordering?

keszybz replied:
It's explained in the commit message. any < template < instance.

poettering commented on src/basic/hash-funcs.c:
```diff
@@ -10,6 +10,9 @@ void string_hash_func(const char *p, struct siphash *state) {
 }
 
 DEFINE_HASH_OPS(string_hash_ops, char, string_hash_func, string_compare_func);
+DEFINE_HASH_OPS_FULL(string_hash_ops_with_freeing,
```
maybe call this string_hash_ops_free_free to follow the hashmap_free_free() niming?

poettering commented on src/basic/path-util.c:
```diff
@@ -336,6 +337,8 @@ char *path_simplify(char *path, PathSimplifyFlags flags) {
          * .//./foo//./bar/. becomes ././foo/./bar/.   (if PATH_SIMPLIFY_KILL_DOT is not specified)
          * .//./foo//./bar/. becomes foo/bar           (if PATH_SIMPLIFY_KILL_DOT is specified)
+         * .//./foo/../bar/. becomes ././bar           (if PATH_SIMPLIFY_KILL_DOUBLE_DOT is specified)
+         * .//./foo/../bar/. becomes bar               (if PATH_SIMPLIFY_KILL_DOT and …_DOUBLE_DOT are specified)
```
uhhh, this is not ok in the general case.

If "foo" is a symlink to "quux/piff" then "foo/../bar" resolves to "quux/bar" quite likely, and not to "/bar"...

What's the usecase for this? We really shouldn't do this for the general case.

keszybz replied:
Duh, I messed this up.

poettering commented on src/shared/unit-file.c:
```diff
@@ -21,3 +22,54 @@ bool unit_type_may_template(UnitType type) {
+        dst = basename(target);
+
+        /* Reject cases where the alias or the target unit have instance names. */
+        r = unit_name_to_instance(src, NULL);
```
maybe add a wrapper like this:

```c
#define unit_name_classify(name) unit_name_to_instance(name, NULL);
```

Otherwise it's really hard to grok what the game played is

poettering commented on src/shared/unit-file.c:
```diff
@@ -21,3 +22,54 @@ bool unit_type_may_template(UnitType type) {
+                return log_notice_errno(r, "%s: not a valid unit name \"%s\": %m", filename, src);
+        if (r == UNIT_NAME_INSTANCE)
+                return log_notice_errno(SYNTHETIC_ERRNO(EINVAL),
+                                        "%s: unit symlink has instance name, rejecting.", filename);
```
hmm, waaa? not sure i grok this? we have symlinks like "gettys.target/serial-getty@ttyS5.service" → "../serial-getty@.service" all the time, so why disallow this?

keszybz replied:
This is only about *alias* symliks, not about symlinks in .wants or such. I'll rename the function to make this clearer.

The general idea was that symlinks like `/etc/systemd/system/serial-getty@ttyS1.service → anything` (apart from masking), are either pointless or invalid, so we should forbid them in general.

poettering replied:
hmm, should it be forbidden though?

I mean i see nothign wrong with having many aliases as long as the instance matches...

i.e. foobar@waldo.service and quux@waldo.service should be fine as aliases

keszybz replied:
Do you think it would be OK to have foobar@waldo.service alias quux@waldo.service, and foobar@zoom.service alias foobar-alias@zoom.service? Or should foobar@waldo.service→quux@waldo.service automatically apply to all instances and foobar@.service should alias quux@.service ?

poettering commented on src/shared/unit-file.c:
```diff
@@ -21,3 +22,54 @@ bool unit_type_may_template(UnitType type) {
                       UNIT_PATH);
 }
+
+int unit_validate_symlink(const char *filename, const char *target) {
```
since this logs (which I think is unexpected for such a library-like function), maybe call this unit_validate_symlink_and_warn()?

poettering commented on src/shared/unit-file.c:
```diff
@@ -21,3 +22,54 @@ bool unit_type_may_template(UnitType type) {
+                                        "%s: symlink target name type \"%s\" does not match source, rejecting.",
+                                        filename, dst);
+
+        t2 = unit_name_to_type(dst);
```
i think src_type and dst_type are perttier names than t1 and t2

poettering commented on src/shared/dropin.c:
```diff
@@ -121,12 +115,7 @@ static int unit_file_find_dir(
-
-        chased = NULL;
-        return 0;
+        return strv_consume(dirs, TAKE_PTR(chased));
```
drops logging in case of oom

poettering commented:
I only reviewed everything up to the last three so far. Maybe merge those ahead of time, so that the other three stand on their own feet?

keszybz commented:
I removed the commits which will need reworking, and repushed the remaining ones. I think all your comments are addressed.

yuwata commented on src/basic/strv.c:
```diff
@@ -889,3 +889,73 @@ int fputstrv(FILE *f, char **l, const char *separator, bool *space) {
+                        free(l);
+                        return r;
+                }
+
```
Not a matter, but I'd like to drop this line break.

yuwata commented on src/basic/strv.c:
```diff
@@ -889,3 +889,73 @@ int fputstrv(FILE *f, char **l, const char *separator, bool *space) {
+                        free(l);
+                        return r;
+                }
+
```
Also here.

yuwata commented on src/basic/strv.c:
```diff
@@ -889,3 +889,73 @@ int fputstrv(FILE *f, char **l, const char *separator, bool *space) {
+        return 0;
+}
+
+int string_strv_ordered_hashmap_put(OrderedHashmap **h, const char *key, const char *value) {
```
Is it possible to merge the two new function by casting to `HashmapBase`?

yuwata commented on src/basic/hashmap.c:
```diff
@@ -1768,6 +1768,32 @@ int set_consume(Set *s, void *value) {
+        if (r < 0)
+                return r;
+
+        _cleanup_free_ char *kdup = NULL, *vdup = NULL;
```
Ugh, I do not like this... Please move this to the beginning of the function.

keszybz replied:
Well, by declaring this here, I'm telling the reader that those variables are not used before.
The compiler doesn't care. Why do you think it is better to move them to the top?

yuwata replied:
I did not say it is better, but I prefer to move them to the beginning. I know we dropped that rule from our CODING_STYLE. So, it is a matter of taste. And this is your PR. I'll push "Resolve conversation" button.

yuwata commented on src/basic/strv.c:
```diff
@@ -900,28 +900,36 @@ int string_strv_hashmap_put(Hashmap **h, const char *key, const char *value) {
         l = hashmap_get(*h, key);
         if (l) {
-                /* A list for this key already exists, let's append to it */
+                /* A list for this key already exists, let's append to it if it is not listed yet */
```
I think this commit should be merged with the previous one.

keszybz commented:
Updated with the suggested changes, except for the variable declaration location. I can move that if necessary, but I think it's better as is. I merged various commits which add new functions and the tests for those functions, I hope that's what you had in mind.

I also added one more trivial cleanup commit.

yuwata commented:
LGTM.
